### PR TITLE
rustdoc: Fix missing private inlining

### LINKED ISSUE #109258: Missing inlining for private re-export
This code is failing:

```rust
pub use crate::bar::Foo;
pub mod bar;

// bar/mod.rs
mod priv_mod;
pub use self::priv_mod::Foo;

// bar/priv_mod.rs
pub struct Foo;
```

`Foo` should be inlined in `bar` but isn't. The regression appeared in https://github.com/rust-lang/rust/pull/108870.

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes https://github.com/rust-lang/rust/issues/109258.

If the item isn't inlined, it shouldn't have been added into `view_item_stack`. The problem here was that it was not removed, preventing sub items to be inlined if they have a re-export in "upper levels".

cc @epage 
r? @notriddle

### CODE REVIEW

File: tests/rustdoc/issue-109258-missing-private-inlining.rs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+// @!has 'foo/struct.Foo.html'
+pub use crate::issue_109258::Foo;
+
+// @has 'foo/issue_109258/index.html'
+// We should only have a "Structs" header.
+// @count - '//*[@id="main-content"]/h2[@class="small-section-header"]' 1
+// @has - '//*[@id="main-content"]/h2[@class="small-section-header"]' 'Structs'
+// @has - '//*[@id="main-content"]//a[@href="struct.Foo.html"]' 'Foo'
+// @has 'foo/issue_109258/struct.Foo.html'
+pub mod issue_109258;
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply tests/rustdoc/issue-109258-missing-private-inlining.rs:22]: I suppose. Let me add a test case for it as well.

**Comment by matthiaskrgr** [comment]
@bors r-
Looks like theres a stack overflow now?
https://github.com/rust-lang-ci/rust/actions/runs/4451938882/jobs/7819150686#step:26:9516

**Comment by GuillaumeGomez** [comment]
It's the only rustdoc PR in the rollup but I can't reproduce the CI failure locally and I don't really see why this test in particular would now fail. Quite surprising... Maybe it's a mix with another PR in the rollup that gave this result?

**Comment by GuillaumeGomez** [comment]
i really can't figure out from my tests how it could make this error. Let's retry then.

@bors r=notriddle

**Pull request merged**